src/Pact/Eval.hs

```diff
@@ -678,41 +672,129 @@ reduceDirect TNative {..} as ai =
-  -- get step from environment or create a new one
-  PactStep{..} <- view eePactStep >>= \ps -> case ps of
-    Nothing -> view eeHash >>= \hsh ->
-      return $ PactStep 0 False (toPactId hsh) Nothing
-    Just v -> return v
+initPact :: Info -> PactContinuation -> Term Ref -> Eval e (Term Name)
+initPact i app bod = view eePactStep >>= \es -> case es of
+  Just v -> evalError i $ "initPact: internal error: step already in environment: " <> pretty v
+  Nothing -> view eeHash >>= \hsh ->
+    applyPact i app bod $ PactStep 0 False (toPactId hsh) Nothing
+
+
+-- | Apply or resume a pactdef step.
+applyPact :: Info -> PactContinuation -> Term Ref -> PactStep -> Eval e (Term Name)
+applyPact i app (TList steps _ _) PactStep {..} = do
+
+  -- only one pact state allowed in a transaction
+  use evalPactExec >>= \bad -> unless (isNothing bad) $
+    evalError i "Multiple or nested pact exec found"
+
   -- retrieve indicated step from code
-  s <- maybe (evalError i $ "applyPact: step not found: " <> pretty _psStep) return $ steps V.!? _psStep
-  case s of
-    step@TStep {} -> do
-      stepEntity <- traverse reduce (_tStepEntity step)
-      let
-        initExec executing = evalPactExec .=
-          Just (PactExec (length steps) Nothing executing _psStep _psPactId app)
-        execStep = do
-          initExec True
-          case (_psRollback,_tStepRollback step) of
-            (False,_) -> reduce $ _tStepExec step
-            (True,Just rexp) -> reduce rexp
-            (True,Nothing) -> return $ tStr $ renderCompactText' $
-              "No rollback on step " <> pretty _psStep
-      case stepEntity of
-        Just (TLitString se) -> view eeEntity >>= \envEnt -> case envEnt of
-          Just (EntityName en) | se == en -> execStep -- matched for "private" step exec
-                               | otherwise -> initExec False >> return (tStr "Skip step")
-          Nothing -> evalError (_tInfo step) "Private step executed against non-private environment"
-        Just t -> evalError (_tInfo t) "step entity must be String value"
-        Nothing -> execStep -- "public" step exec
+  st <- maybe (evalError i $ "applyPact: step not found: " <> pretty _psStep) return $ steps V.!? _psStep
+  step <- case st of
+    TStep step _i -> return step
     t -> evalError (_tInfo t) "expected step"
-applyPact _ t = evalError (_tInfo t) "applyPact: expected list of steps"
+
+  -- determine if step is skipped (for private execution)
+  (executing,private) <- traverse reduce (_sEntity step) >>= \stepEntity -> case stepEntity of
+    Nothing -> return (True,False)
+    Just (TLitString se) -> view eeEntity >>= \envEnt -> case envEnt of
+      Just (EntityName en) -> return (se == en,True) -- execute if req entity matches context entity
+      Nothing -> evalError' step "applyPact: private step executed against non-private environment"
+    Just t -> evalError' t "applyPact: step entity must be String value"
```

Comment:
If this is what we're using `_sEntity` for, then we could probably make it a `Maybe EntityName` so we can avoid reducing and then pat matching. But maybe I'm misunderstanding what `_sEntity` is. Could you explain why it's a term?

Reply:
In the code 
```
(defpact foo (sender receiver bar baz)
  (step sender (quux bar))
  (step receiver (go baz)))
```
`step sender` ensures the first step will only be executed by nodes bearing the value of `sender` as their entity name, and likewise for receiver. As variables they must be a `Term`.

Reply:
But there should be some tests around this, up until now privacy-oriented code hasn't lived in Pact but in ScalableBFT, so this is totally unexercised/uncovered :\